components/compliance-service/integration_test/stats_server_read_trend_test.go

```diff
@@ -0,0 +1,187 @@
+	for k, v := range reportsProjects {
+		projectRules[k] = &iam_v2.ProjectRules{
+			Rules: []*iam_v2.ProjectRule{
+				{
+					Conditions: []*iam_v2.Condition{
+						{
+							Type:   iam_v2.ProjectRuleConditionTypes_ROLES,
+							Values: v,
+						},
+					},
+				},
+			},
+		}
+	}
+
+	// Send a project rules update event
+	esJobID, err := suite.ingesticESClient.UpdateReportProjectsTags(everythingCtx, projectRules)
+	assert.Nil(t, err)
+
+	suite.WaitForESJobToComplete(esJobID)
+
+	suite.RefreshComplianceReportIndex()
+
+	octoberTwentyFifthQuery := stats.Query{
+		Filters: []*stats.ListFilter{
+			{Type: "start_time", Values: []string{"2018-10-24T23:59:59Z"}},
+			{Type: "end_time", Values: []string{"2018-10-25T23:59:59Z"}},
+		},
+	}
+	successCases := []struct {
+		description     string
+		allowedProjects []string
+		query           stats.Query
+
+		expectedPassedCnt int32
+	}{
+		{
+			description:     "Projects: user has access to all projects",
+			allowedProjects: []string{authzConstants.AllProjectsExternalID},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 5,
+		},
+		{
+			description:     "Projects: user has access to one project with reports",
+			allowedProjects: []string{"project1"},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 2,
+		},
+		{
+			description:     "Projects: user has access to some projects with reports",
+			allowedProjects: []string{"project1", "project2"},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 4,
+		},
+		{
+			description:     "Projects: user has access to projects without reports",
+			allowedProjects: []string{"project4", "project5"},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 0,
+		},
+		{
+			description:     "Projects: user has access to one project with reports and unassigned reports",
+			allowedProjects: []string{"project1", authzConstants.UnassignedProjectID},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 3,
+		},
+		{
+			description:     "Projects: user has access to some projects with reports and unassigned reports",
+			allowedProjects: []string{"project1", "project2", authzConstants.UnassignedProjectID},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 5,
+		},
+		{
+			description:     "Projects: user has access to projects without reports and unassigned reports",
+			allowedProjects: []string{"project4", "project5", authzConstants.UnassignedProjectID},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 1,
+		},
+		{
+			description:     "Projects: user has access to unassigned reports",
+			allowedProjects: []string{authzConstants.UnassignedProjectID},
+			query:           octoberTwentyFifthQuery,
+
+			expectedPassedCnt: 1,
+		},
+	}
+
+	for _, test := range successCases {
+		t.Run(test.description, func(t *testing.T) {
+			ctx := contextWithProjects(test.allowedProjects)
+
+			test.query.Type = "nodes"
```

Comment:
Is there a reason this is not placed in the creation of the query at line 94? The same with the line 177 `test.query.Type = "controls"`

Reply:
Yes.  The reason these are set late like this is because they don't coexist for each call.
I set Type to "nodes" first, then make the call to the backend "ReadTrend" api. This returns node oriented trend data.  I then make appropriate assertions and then move on.  The next test is with type "control".. again, making backend call with "control" set as Type.  

I could have, instead, had a separate block dedicated to each respective type, but there would have been a lot of duplicate code.. so I combined.

Reply:
That makes sense. Thanks